src/puya/ir/main.py

```diff
@@ -146,10 +173,29 @@ def _expand_tuple_params(args: Sequence[awst_nodes.SubroutineArgument]) -> Itera
                     atype=wtype_to_avm_type(arg.wtype),
                 )
             )
-            idx += 1
 
 
-def _make_subroutine(func: awst_nodes.Function) -> Subroutine:
+def _should_include_implicit_returns(func: awst_nodes.Function, ctx: IRBuildContext) -> bool:
+    """
+    Determine if a function should implicitly return mutable reference parameters.
+
+    ABI methods which are only called by the ABI router in the approval_program do not need to
+    implicitly return anything as we know our router is not interested in anything but the explicit
+    return value.
+
+    Anything else would require further analysis, so err on the side of caution and include include
+    the implicit returns.
+    """
+    if not isinstance(func, awst_nodes.ContractMethod) or not func.abimethod_config:
+        return True
+    return any(
+        f
+        for f in ctx.function_call_sites.get(func, [])
+        if isinstance(f, awst_nodes.ContractMethod) and f.name != "approval_program"
```

Comment:
hmmm the string comparison here is a bit tenuous but I can see why you've done it that way

Reply:
is there a more robust way to identify an approval_program other than by name? I guess we need to know if the approval_program was generated by puya, or manually too.

Reply:
Yeah it's not particularly easy currently especially with folding, we might need to do a slight refactor there to expose that neatly, I'd be okay if we merged this as is - maybe add a `REFACTOR` comment :P